client/Client.hs

```diff
@@ -1,41 +1,157 @@
-            Left s -> do
-                error . message $ s
-        Nothing -> do
-          if retryCount < 1
-            then error "retry count exceeded, no message recieved"
-          else do
-            threadDelay 1000000
-            connect host port $ retryCount - 1
+  sockLock <- newTVarIO sock
+  configLock <- newEmptyTMVarIO
+  routerLock <- newTVarIO Map.empty
+  bufferLock <- newTVarIO BS.empty
+  let nats = Nats sockLock routerLock configLock bufferLock
+
+  forkIO . forever $ readSock nats
+  connect' nats
+  forkIO . forever $ loop nats
+  return nats
+
+connect' :: Nats -> IO ()
+connect' nats = do
+  -- TODO: at some point we'll want to read the state set by INFO
+  writeSock nats $ Connect {
+    Types.Connect.verbose       = False,
+    Types.Connect.pedantic      = True,
+    Types.Connect.tls_required  = False,
+    Types.Connect.auth_token    = Nothing,
+    Types.Connect.user          = Nothing,
+    Types.Connect.pass          = Nothing,
+    Types.Connect.name          = Nothing,
+    Types.Connect.lang          = "Haskell",
+    Types.Connect.version       = "2010",
+    Types.Connect.protocol      = Nothing,
+    Types.Connect.echo          = Nothing,
+    Types.Connect.sig           = Nothing,
+    Types.Connect.jwt           = Nothing,
+    Types.Connect.no_responders = Nothing,
+    Types.Connect.headers       = Nothing
+  }
+
+pong :: Nats -> IO ()
+pong nats = writeSock nats Pong
+
+pub :: Nats -> BS.ByteString -> BS.ByteString -> IO ()
+pub nats subject payload = writeSock nats $ Pub subject Nothing Nothing (Just payload)
+
+sub :: Nats -> BS.ByteString -> BS.ByteString -> (Msg -> IO()) -> IO ()
+sub nats sid subject callback = do
+  atomically . modifyTVar (router nats) $ \m -> Map.insert sid callback m
+  writeSock nats $ Sub subject Nothing sid
+
+unsub :: Nats -> BS.ByteString -> BS.ByteString -> IO ()
+unsub nats sid subject = do
+  atomically . modifyTVar (router nats) $ \m -> Map.delete subject m
+  writeSock nats $ Sub subject Nothing sid
+
+readSock :: Nats -> IO ()
+readSock nats = do
+  socket <- readTVarIO (sock nats)
+  dat    <- TCP.recv socket 1024
+  case dat of
+    Nothing  -> do
+      -- we've reached the end of input
+      threadDelay 1000000
+      readSock nats
+    Just msg -> void . forkIO $ writeToBuffer nats msg
+
+writeSock :: (Transformer a, Validator a) => Nats -> a -> IO ()
+writeSock nats msg = do
+  case validate msg of
+    Left err -> error $ show err
+    Right _  -> do
+      socket <- readTVarIO (sock nats)
+      TCP.send socket $ transform msg
+
+loop :: Nats -> IO ()
+loop nats = do
+  msg <- atomically $ readBuffer nats
+  case msg of
+    ParsedMsg a     -> handleMsg nats a
+    ParsedPing Ping -> pong nats
+    ParsedInfo a    -> setConfigFromInfo nats a
+    ParsedOk _      -> return ()
+    ParsedPong _    -> return ()
+    -- TODO: we should check the error to see if it' fatal, if so we'll have been disconnected
```

Comment:
Raise an issue to write this logic into a state machine, there are a couple of states that we want to deal with

Reply:
#113